src/main/java/org/kiwiproject/consul/util/failover/strategy/BlacklistingConsulFailoverStrategy.java

```diff
@@ -91,43 +82,75 @@ public Optional<Request> computeNextStage(@NonNull Request previousRequest, @Nul
      */
     private Optional<HostAndPort> findTargetNotInBlacklist() {
         return targets.stream().filter(target -> {
+            if (isNotBlacklisted(target)) {
+                return true;
+            }
 
-            // If we have blacklisted this key
-            if (blacklist.containsKey(target)) {
-
-                // Get when we blacklisted this key
-                final Instant blacklistedAt = blacklist.get(target);
-
-                // If the duration between the blacklist time ("then") and "now" is greater than the
-                // timeout duration (Duration(then, now) - timeout < 0), then remove the blacklist entry
-                if (Duration.between(blacklistedAt, Instant.now()).minusMillis(timeout).isNegative()) {
-                    return false;
-                } else {
-                    blacklist.remove(target);
-                    return true;
-                }
-            } else {
+            if (isPastBlacklistDuration(target)) {
+                // Remove the target from blacklist once timeout has expired
+                blacklist.remove(target);
                 return true;
             }
+
+            return false;
         }).findAny();
     }
 
+    @VisibleForTesting
+    boolean isPastBlacklistDuration(HostAndPort target) {
+        var blacklistedAt = blacklist.get(target);
+
+        // Exit early if the target is not actually in the blacklist.
+        // The only way this can be true is if a concurrent request causes the
+        // target to be removed from the blacklist between the time it is first
+        // checked in #findTargetNotInBlacklist and when this method gets the
+        // value of the target key above. If that happens, just return true. It
+        // won't cause any problems with the Map#remove operation since that
+        // will just return null if there was no mapping with the given key.
+        if (isNull(blacklistedAt)) {
+            return true;
+        }
+
+        // If the duration between the blacklist time ("then") and "now" is greater than the
+        // timeout duration (Duration(then, now) - timeout >= 0), then the timeout has passed
+        var adjustedBlacklistDuration = Duration.between(blacklistedAt, Instant.now()).minusMillis(timeout);
+
+        return isPositiveOrZero(adjustedBlacklistDuration);
+    }
+
+    private static boolean isPositiveOrZero(Duration duration) {
+        return !duration.isNegative();
+    }
+
     private static boolean previousResponseFailedAndWasNot404(Response previousResponse) {
         return nonNull(previousResponse) && !previousResponse.isSuccessful() && previousResponse.code() != 404;
     }
 
     @Override
     public boolean isRequestViable(@NonNull Request current) {
-        return (targets.size() > blacklist.size()) || !blacklist.containsKey(fromRequest(current));
+        return findTargetNotInBlacklist().isPresent();
```

Comment:
the check is for the `current` request, but this iterates through the entire `targets`, so the `current` parameter is misleading -- method/signature change?

also, this has a minor performance penalty since it could loop through the list.

Reply:
Agree that the parameter is misleading based on my reading of the rest of the Javadoc, which says:

> Determines if there is a viable candidate for the next request

And the `@return` description says:

> A boolean representing if there is another possible request candidate available.

So, it seems like this method really is meant to check whether there is _any_ target available to handle it. And so I think the parameter should be changed to just the generic `request`. And also probably for the `markRequestFailed` method as well, for the same reason.

As for the implementation, if you do not check other if targets are available, then once all targets are  in the blacklist, they don't ever come out of it - i.e., the `CheckConsul` program you provided tries `a.foo.bar.baz` and `b.foo.bar.baz` on the first attempt, and after that `ConsulFailoverInterceptor` always throws an exception with the message `"Consul failover strategy has determined that there are no viable hosts remaining"`. 

But I agree maybe we can first do the original checks (are there more targets than in blacklist, or is the current request target not in blacklist) which keeps things faster. And then if those both fail, we tack on the `findTargetNotInBlacklist().isPresent()` check, which may have the side-effect of removing a target from the blacklist. The code becomes:

```java
    @Override
    public boolean isRequestViable(@NonNull Request request) {
        return atLeastOneTargetIsAvailable() ||
                isNotBlacklisted(hostAndPortFromRequest(request)) ||
                findTargetNotInBlacklist().isPresent();
    }

    private boolean atLeastOneTargetIsAvailable() {
        return targets.size() > blacklist.size();
    }
```

Another option would be to use a background thread that periodically executes. It would check whether any targets have passed their expiration time in the blacklist, and remove them. But, even though I don't really love the side-effect in `findTargetNotInBlacklist`, I also don't really think it's worth having a blacklist "pruner" thread. `findTargetNotInBlacklist` is private so outside client code won't be calling it directly.

Reply:
In terms of functionality, it looks fine to me.

What is misleading is the method name `isRequestViable(@NonNull Request ...)` implies checking _the_ request to see if it is viable, instead of _any_ request that is viable.

I'm not quite sure why `isRequestViable` is needed -- can it let `computeNextStage` run until all hosts are exhausted?

> But, even though I don't really love the side-effect in findTargetNotInBlacklist, I also don't really think it's worth having a blacklist "pruner" thread.

Yep, agreed.  The side-effect is unexpected.